Sources/DataPipeline/DataPipelineImplementation.swift

```diff
@@ -76,11 +81,49 @@ class DataPipelineImplementation: DataPipelineInstance {
     var profileAttributes: [String: Any] {
         get { analytics.traits() ?? [:] }
         set {
-            let userId = analytics.userId ?? analytics.anonymousId
-            analytics.identify(userId: userId, traits: newValue)
+            let userId = analytics.userId
+            guard let userId = userId else {
+                logger.error("No user identified. If you don't have a userId but want to record traits, please pass traits using identify(body: Codable)")
+                return
+            }
+            commonIdentifyProfile(userId: userId, attributesDict: newValue)
         }
     }
 
+    private func commonIdentifyProfile(userId: String, attributesDict: [String: Any]? = nil, attributesCodable: Codable? = nil) {
+        let currentlyIdentifiedProfile = analytics.userId
+        let isChangingIdentifiedProfile = currentlyIdentifiedProfile != nil && currentlyIdentifiedProfile != userId
+        let isFirstTimeIdentifying = currentlyIdentifiedProfile == nil
+
+        if isFirstTimeIdentifying || isChangingIdentifiedProfile {
+            // logger.debug("running hooks profile identified \(userId)")
+            // FIXME: [CDP] Request Journeys to invoke profile identify hooks
+            // hooks.profileIdentifyHooks.forEach { hook in
+            //     hook.profileIdentified(identifier: userId)
+            // }
+        }
+        if let attributes = attributesCodable {
+            analytics.identify(userId: userId, traits: attributes)
+        } else {
+            analytics.identify(userId: userId, traits: attributesDict)
+        }
+    }
+
+    private func commonClearIdentify() {
+        // logger.debug("deleting device info from \(currentlyIdentifiedProfile) to stop sending push to a profile that is no longer identified")
+        // TODO: [CDP] Confirm how can we delete devices for CDP
+
+        // logger.debug("running hooks: profile stopped being identified \(currentlyIdentifiedProfile)")
+        // FIXME: [CDP] Request Journeys to invoke profile clearing hooks
+        // hooks.profileIdentifyHooks.forEach { hook in
+        //     hook.beforeProfileStoppedBeingIdentified(oldIdentifier: currentlyIdentifiedProfileIdentifier)
+        // }
```

Comment:
similar comment as with profile identified hooks --> registering device tokens. Maybe we don't need hooks/journeys at all here? Maybe this function can internally delete device token from profile?